source/Halibut/Transport/Protocol/WebSocketStream.cs

```diff
@@ -37,47 +40,74 @@ public override void SetLength(long value)
         public override int Read(byte[] buffer, int offset, int count)
         {
             AssertCanReadOrWrite();
-            var segment = new ArraySegment<byte>(buffer, offset, count);
-            var receiveResult = context.ReceiveAsync(segment, CancellationToken.None)
+            return ReadAsync(buffer, offset, count, CancellationToken.None)
                 .ConfigureAwait(false).GetAwaiter().GetResult();
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            AssertCanReadOrWrite();
+            var segment = new ArraySegment<byte>(buffer, offset, count);
+            var receiveResult = await context.ReceiveAsync(segment, cancellationToken);
             return receiveResult.Count;
         }
 
-        public async Task<string> ReadTextMessage()
+        public async Task<string> ReadTextMessage(CancellationToken cancellationToken)
```

Comment:
This method should not be changed. We only want to  change the implementation if Async Halibut is enabled

Reply:
So here I would add a new method `public async Task<string> ReadTextMessageAsync(CancellationToken cancellationToken)` that has your new implementation and leave the old one alone